checker/tests/lock/TestTreeKinds.java

```diff
@@ -331,65 +487,131 @@ public void testMethodAnnotations() {
 
   if (r.nextBoolean()) {
     lockTheLock();
-    // Yes, lock reassignments are bad, but for the purposes of this
-    // experiment they are useful.
-    @LockHeld ReentrantLock heldLock = lock;
-    foo = new Object();
+    foo.field.toString();
 
     unlockTheLock();
 
-    //:: error: (assignment.type.incompatible)
-    @LockHeld ReentrantLock heldLock2 = lock;
     //:: error: (contracts.precondition.not.satisfied.field)
-    foo = new Object();
+    foo.field.toString();
   }
   else {
-    //:: error: (assignment.type.incompatible)
-    @LockHeld ReentrantLock heldLock = lock;
     //:: error: (contracts.precondition.not.satisfied.field)
-    foo = new Object();
+    foo.field.toString();
   }
 
   if (tryToLockTheLock()) {
-    @LockHeld ReentrantLock heldLock = lock;
-    foo = new Object();
+    foo.field.toString();
   }
   else {
-    //:: error: (assignment.type.incompatible)
-    @LockHeld ReentrantLock heldLock = lock;
     //:: error: (contracts.precondition.not.satisfied.field)
-    foo = new Object();
+    foo.field.toString();
   }
 
+  // TODO: An error is raised here because a non-side-effect-free method is called above,
```

Comment:
Fixed the comment.
`nextBoolean` is defaulted to @ReleasesNoLocks, which while not being @SideEffectFree, is side effect free for the purposes of resetting the estimate of locks that are held.
Methods with side effects are soundly handled but too conservatively. This could be improved by detecting which local variables are definitely not affected by a method call. However this would require tracking of aliases and leaks:

    final ReentrantLock lock = new ReentrantLock();
    someObject.methodThatReleasesNoLocks(lock);
    someObject.methodThatMayReleaseLocks();
    // lock is not known to be held at this point

In the call to `methodThatReleasesNoLocks`, a reference to `lock` may have been stored in `someObject`, and `methodThatMayReleaseLocks()` may release it.